src/routes/ui.rs

```diff
@@ -1,15 +1,168 @@
+                input type="password" id="password" name="password";
+            }
+
+            button type="submit" "sign in"
+        }
+    })
+}
+
+#[derive(Debug, FromForm)]
+pub struct SigninForm {
+    username: String,
+    password: String,
+}
+
+#[post("/auth/sign_in", data = "<form>")]
+pub fn auth_signin_post(
+    form: Form<SigninForm>,
+    mut cookies: Cookies,
+    db_conn: db::Connection,
+) -> Result<Flash<Redirect>, Error> {
+    let form_data = form.get();
+    let user = User::by_username(&db_conn, &form_data.username)?;
+
+    if let Some(user) = user {
+        if user.valid_password(&form_data.password) {
+            cookies.add_private(Cookie::new("uid", user.id.to_string()));
+
+            return Ok(Flash::success(Redirect::to("/"), "signed in!"));
+        }
+    }
+
+    Ok(Flash::error(
+        Redirect::to("/auth/sign_in"),
+        "wrong password (or username!)",
+    ))
+}
+
+#[post("/auth/sign_out")]
+pub fn auth_signout(user: Option<User>, mut cookies: Cookies) -> Redirect {
+    if user.is_some() {
+        cookies.remove_private(Cookie::named("uid"));
+    }
+
+    Redirect::to("/")
+}
+
+#[derive(FromForm, Validate, Debug)]
+pub struct SignupForm {
+    #[validate(length(min = "1", max = "32"))]
+    username: String,
+    #[validate(email)]
+    email: String,
+    #[validate(length(min = "3", max = "64",
+                      message = "Password must be between 3 and 64 characters long."))]
+    password: String,
+}
+
+#[get("/auth/sign_up")]
+pub fn auth_signup_get(flash: Option<FlashMessage>) -> Page {
+    Page::new().title("sign up").flash(flash).content(html! {
+        header h2 "sign up"
+
+        form method="post" {
+            div {
+                label for="username" "username:"
+                input type="text" id="username" name="username";
+            }
+
+            div {
+                label for="email" "email:"
+                input type="email" id="email" name="email";
+            }
+
+            div {
+                label for="password" "password:"
+                input type="password" id="password" name="password";
+            }
+
+            button type="submit" "sign up"
+        }
+    })
+}
+
+#[post("/auth/sign_up", data = "<form>")]
+pub fn auth_signup_post(
+    form: Form<SignupForm>,
+    db_conn: db::Connection,
+) -> Result<Flash<Redirect>, Error> {
+    let form_data = form.get();
+
+    if let Err(errs) = form_data.validate() {
+        let errs = errs.inner();
+
+        // concatenate the error descriptions, with commas between them.
+        // TODO: make this less ugly :(
+        let error_desc = errs.iter()
+            .flat_map(|(_, errs)| errs)
+            .map(|e| {
+                let msg = e.message.to_owned();
+                msg.unwrap_or(Cow::Borrowed("unknown error"))
+            })
+            .join(", ");
+
+        return Ok(Flash::error(Redirect::to("/auth/sign_up"), error_desc));
+    }
+
+    let account = NewAccount {
```

Comment:
It seems to me that creating a user and account is conceptually a single change to the database.  If this is true, this should be wrapped up in a transaction.  Maybe something like

```rust
(*db_conn).transaction::<_, Error, _>(|| {
  let account = ...;
  let user = ...;
  Ok(Flash::error(...))
})
```

To the best of my knowledge, doing this will require re-exporting the `diesel::connection::Connection` trait from the `db` crate (under some other name like `DieselConn`: we already use `Connection`).  Another possibility would be to make `db::Connection` implement `diesel::connection::Connection`, but I'm not sure how much work that is.

Reply:
OH YEAH WHOOPS THANKS

:sweat:

Reply:
Rather than reexporting the trait, we could also depend on Diesel in the main crate, although that might increase build times (and CI times, depending on how aggressive Travis is :stuck_out_tongue:)

Reply:
Implemented the reexport thing, will test and push when I get back to wifi